src/swapRouter.ts

```diff
@@ -300,14 +336,91 @@ export abstract class SwapRouter {
       }
     }
 
-    // refund
-    if (mustRefund) {
+    // must refund when paying in ETH, but with an uncertain input amount
+    if (inputIsNative && sampleTrade.tradeType === TradeType.EXACT_OUTPUT) {
       calldatas.push(Payments.encodeRefundETH())
     }
 
     return {
       calldata: MulticallExtended.encodeMulticall(calldatas, options.deadlineOrPreviousBlockhash),
-      value: toHex(totalValue.quotient),
+      value: toHex(inputIsNative ? totalAmountIn.quotient : ZERO),
+    }
+  }
+
+  /**
+   * Produces the on-chain method name to call and the hex encoded parameters to pass as arguments for a given trade.
+   * @param trade to produce call parameters for
+   * @param options options for the call parameters
+   */
+  public static swapAndAddCallParameters(
+    trades:
+      | Trade<Currency, Currency, TradeType>
+      | V2Trade<Currency, Currency, TradeType>
+      | V3Trade<Currency, Currency, TradeType>
+      | (V2Trade<Currency, Currency, TradeType> | V3Trade<Currency, Currency, TradeType>)[],
+    options: SwapOptions,
+    tokenInBalanceDesired: CurrencyAmount<Currency>,
+    tokenOutBalanceDesired: CurrencyAmount<Currency>,
+    position: Position
+  ): MethodParameters {
+    invariant(!!options.addLiquidityOptions, 'MISSING_ADD_LIQUIDITY_OPTIONS')
+
+    const {
+      calldatas,
+      inputIsNative,
+      outputIsNative,
+      totalAmountIn: totalAmountSwapped,
+      totalAmountOut,
+    } = SwapRouter.encodeSwaps(trades, options)
+
+    const totalAmountIn = totalAmountSwapped.add(tokenInBalanceDesired)
+
+    // if tokens are native they will be converted to WETH9
+    const tokenIn = inputIsNative ? WETH9[1] : tokenInBalanceDesired.currency.wrapped
+    const tokenOut = outputIsNative ? WETH9[1] : tokenOutBalanceDesired.currency.wrapped
+
+    // if swap output does not make up whole outputTokenBalanceDesired, pull in remaining tokens for adding liquidity
+    const amountOutRemainingBalance = tokenOutBalanceDesired.subtract(totalAmountOut)
+    if (amountOutRemainingBalance.greaterThan(CurrencyAmount.fromRawAmount(tokenOutBalanceDesired.currency, 0))) {
+      // if output is native, convert value to WETH9, else pull ERC20 token
+      outputIsNative
+        ? calldatas.push(
+            SwapRouter.INTERFACE.encodeFunctionData('pay', [
```

Comment:
wow, interesting, never would have thought to use `pay` to wrap ETH into WETH like this, but i guess it works? if we end up doing it this way, deserves a special-purpose function in the v3-sdk `payments.ts` though.

however, i think we should probably add a special-purpose smart contract function for this use case, because the `payer` argument isn't doing anything here, and also we're triggering an extraneous `IWETH9(WETH9).transfer(recipient, value);` call.

wdyt about a `PeripheryPaymentsExtended#wrapETH` or similar?

Reply:
yup, I think we need a wrapETH function!